%% GENERIC STYLE SETTINGS
\input{sections/preamble_style}
% LISTINGS SETTING
\input{sections/preamble_codelistings}

%%%
% TITLE PREAMBLE
\title[Open science and Bio* projects] % (optional, only for long titles)
{Open science and Bio* projects}
\subtitle{Standing on each other's shoulders}
\author[Cock] % (optional, for multiple authors)
{Peter~Cock}
\institute[The James Hutton Institute] % (optional)
{
  Information and Computational Sciences\\
  The James Hutton Institute
}
\date[July 2017] % (optional)
{EASTBIO Genomic Approaches, 1$^{st}$ July 2017}
\subject{Bioinformatics}

%%%
% TOC
\input{sections/preamble_toc}

%%%
% START DOCUMENT
\begin{document}

\frame[plain]{\titlepage}

\section{Introduction}
%\subsection{BLAST Extension}

\begin{frame}
  \frametitle{Progress in Science}
  \begin{itemize}
    \item \emph{``If I have seen further it is by standing on the shoulders of Giants.''} (Issac Newton, 1676)
    \item Science works best when can build on past work
    \item Methods section in papers \emph{should} let you do this
    \item For computation methods, would ideally get the code...
  \end{itemize}
\end{frame}

\begin{frame}
  \frametitle{Progress in Open Source}
  \begin{itemize}
    \item By default, computer code is under copyright:
      \begin{itemize}
        \item YOU NEED PERMISSION to use, re-use, adapt.
      \end{itemize}
    \item Open Source licensing defines explicit terms:
      \begin{itemize}
        \item GIVES PERMISSION to use, re-use, share, change, at no cost.
        \item That includes packaging for easier installation
      \end{itemize}
  \end{itemize}
\end{frame}

\section{Open Source}

\subsection{Open Source Licenses}

\begin{frame}
  \frametitle{Many Open Source Licenses}
  \begin{itemize}
    \item Open Source Initiative \url{https://opensource.org/} has a list
    \item Very liberal, e.g. MIT and BSD
    \item `Viral', e.g. GNU Public Licence (GPL)
  \end{itemize}
  Note: Combinging code from multiple  sources can hit license problems
\end{frame}

\subsection{Code sharing}

\begin{frame}
  \frametitle{Code sharing / development}
  Most projects use a web-based platform:
  \begin{itemize}
    \item \href{https://bitbucket.org/product}{Bitbucket}
    \item \href{https://github.com/}{GitHub}
    \item \href{https://about.gitlab.com/}{GitLab}
    \item Sourceforge (popular 10 years ago)
    \item Lab specific website
  \end{itemize}
\end{frame}

\begin{frame}
  \frametitle{GitHub/BitBucket's Business Model}
  \begin{itemize}
    \item Offer their services for free to \emph{public} projects
    \item Offer private projects for fee to \emph{students}
    \item Charge \$\$\$ for private projects
  \end{itemize}
  i.e. We're freeloading and can't count of this continuing.

  Compare with SourceForge, which is now covered in adverts
\end{frame}

\subsection{Open Source in Biology}

\begin{frame}
  \frametitle{Open Source in Biology}
  \begin{itemize}
    \item Much early bioinformatics by Systems Administrators
    \item e.g. \emph{How Perl Saved the Human Genome Project} (Lincoln Stein, 1996)
    \item Absorbed ideas/culture from free and open source world, e.g. Linux, Perl
    \item Open source required championing in early days, e.g. \href{https://www.open-bio.org/wiki/BOSC}{Bioinformatics Open Source Conference} since 2000.\footnote{Peter is a former co-chair}
    \item Many journals now on-board, e.g.
    \href{http://journals.plos.org/plosone/s/materials-and-software-sharing}{PLOS},
    \href{https://academic.oup.com/nar/pages/Policies}{NAR},
    \href{https://academic.oup.com/gigascience/pages/editorial_policies_and_reporting_standards}{GigaScience}
  \end{itemize}
\end{frame}

\begin{frame}
  \frametitle{Open Source Bioinformatics examples}
  Important funded projects in genomics:
  \begin{itemize}
    \item \url{https://github.com/samtools/samtools} \\ Wellcome Trust Sanger Institute
    \item \url{https://github.com/broadinstitute/gatk} \\ Broad Institute
    \item \url{https://github.com/galaxyproject/galaxy} \\ Penn State
  \end{itemize}
\end{frame}

\begin{frame}
  \frametitle{Open Source Bioinformatics examples}
  Large long-running un-funded projects:
  \begin{itemize}
    \item \url{https://github.com/bioperl/bioperl-live} - \href{http://bioperl.org}{BioPerl}
    \item \url{https://github.com/biopython/biopython} - \href{http://biopython.org}{Biopython}\footnote{Peter is a major contributor}
    \item \url{https://github.com/bioruby/bioruby} - \href{http://bioruby.org}{BioRuby}
    \item \url{https://github.com/biojava/biojava} - \href{http://biojava.org}{BioJava}
  \end{itemize}
\end{frame}

\begin{frame}
  \frametitle{Open Source Bioinformatics examples}
  Relevant smaller projects, by individuals or smaller groups:
  \begin{itemize}
    \item \url{https://github.com/arq5x/bedtools}
    \item \url{https://github.com/arq5x/poretools}
    \item \url{https://sourceforge.net/projects/mira-assembler/}
    \item \url{https://github.com/tseemann/prokka}
    \item \url{https://github.com/COMBINE-lab/salmon}
  \end{itemize}
\end{frame}

\begin{frame}
  \frametitle{Open Source Bioinformatics examples}
  Sequencing companies with open source projects:
  \begin{itemize}
    \item \url{https://github.com/illumina} \\ Illumina
    \item \url{https://github.com/nanoporetech/} \\ Oxford Nanopore Technologies
    \item \url{https://github.com/PacificBiosciences} \\ Pacific Biosciences
    \item \url{https://github.com/iontorrent} \\ Ion Torrent
  \end{itemize}
\end{frame}

\begin{frame}
  \frametitle{GitHub is currently very popular}
  \begin{itemize}
    \item Has built up a network effect, the place to go
    \item Continues to innovate and improve the platform
    \item Hub of an ecosystem with other services (e.g. testing)
    \item Also being used for teaching materials, slides, etc
    \item Doubles as an online profile for individuals
    \item Thus far, free for open projects...
  \end{itemize}
\end{frame}

\begin{frame}
  \frametitle{Publishing Source code for long term?}
  \begin{itemize}
    \item Cannot count on GitHub/Bitbucket for long term storage
    \item Personal webpages even more fragile
    \item Scientific record needs long term static archive:
      \begin{itemize}
        \item Journal paper supplementary material
        \item \href{http://about.zenodo.org/}{Zenodo} (powered by CERN)
      \end{itemize}
  \end{itemize}
\end{frame}

\section{Open Data}

\begin{frame}
  \frametitle{Open Data in Genomics}
  Sequence data is \emph{usually} shared, now expected:
  \begin{itemize}
    \item DNA, RNA, Proteins at EMBL-EBI, NCBI, DDBJ
    \item Raw sequencing data too at EMBL SRA, NCBI SRA
    \item Structures at RSCB Protein Data Bank
  \end{itemize}
  Most journals will enforce this.
\end{frame}

\begin{frame}
  \frametitle{Open Data in Genomics}
  Curated data is not always freely openly shared,
\begin{columns}[T]
\begin{column}{.5\textwidth}
%\begin{block}{Image}
  \begin{itemize}
    \item Kyoto Encyclopedia of Genes and Genomes (KEGG) now has \href{http://www.kegg.jp/kegg/legal.html}{restrictions}
    \item The Arabidopsis Information Resource (TAIR) now has \href{https://www.arabidopsis.org/doc/about/tair_subscriptions/413}{fees}
    \item Also some groups like to hoard their precious data...
  \end{itemize}
%\end{block}
\end{column}
\begin{column}{.5\textwidth}
%\begin{block}{Text}
  \includegraphics[width=\textwidth]{Gollum.jpg}
%\end{block}
\end{column}
\end{columns}
\end{frame}

\section{Open Data \& Software $\rightarrow$ Reproducible?}

\begin{frame}
  \frametitle{Open Data \& Software $\rightarrow$ Reproducible?}
  Given a well written scientific paper:
  \begin{itemize}
    \item The raw data will be available (e.g. NCBI accessions)
    \item The tools used and exact versions will be stated
    \item The actual workflow, scripts or command lines are given
  \end{itemize}
  How easy would it be to repeat the analysis?
\end{frame}


\begin{frame}
  \frametitle{Open Data \& Software $\rightarrow$ Reproducible?}
  \begin{itemize}
    \item Sadly click-and-repeat methods in papers are still an aspiration
    \item Being worked on, still lots of technical hurdles (e.g. installing dependencies)
    \item However, openly sharing the data and code is first step!
  \end{itemize}
\end{frame}

\section{Conclusions}

\subsection{Sources of advice}

\begin{frame}
  \frametitle{Sources of coding / tool advice}
  \begin{itemize}
    \item Ask us your peers / supervisor in person
    \item For a given tool check its documentation, mailing lists, issue tracker, online chat/forum, etc
    \item BioStars \url{https://www.biostars.org}
    \item Bioinformatics StackExchange \url{https://bioinformatics.stackexchange.com/}
    \item SeqAnswers Forums \url{http://seqanswers.com/forums}
 \end{itemize}
\end{frame}

\begin{frame}
  \frametitle{Conclusions}
  \begin{itemize}
    \item Open science is a community - help each other
    \item Register a GitHub account to contribute back
      \begin{itemize}
        \item Raise quries; log bugs; offer improvements
        \item Use this to share your own scripts/tools
      \end{itemize}
    \item Your GitHub (etc) profile can be part of your CV, e.g.
      \begin{itemize}
        \item \url{https://github.com/peterjc} - my GitHub profile
        \item \url{https://bitbucket.org/peterjc/} - my BitBucket profile
        \item \url{https://gitlab.com/pjacock} - my GitLab profile
        \item \url{https://www.biostars.org/u/146/} - my BioStars profile
      \end{itemize}
 \end{itemize}
\end{frame}

\subsection{Further reading}

\begin{frame}
  \frametitle{Further reading}
  \begin{itemize}

    \item Mike Croucher \emph{Is your research software correct?}
    \url{https://mikecroucher.github.io/MLPM_talk/}
    % inc. Croucher's Law \emph{I can be an idiot and will make mistakes}

    \item Software Sustainability Institute
    \url{https://www.software.ac.uk/}

    \item Software Carpentry
    \url{https://software-carpentry.org/}

    \item Wilson et al (2014) Best Practices for Scientific Computing
    \url{https://doi.org/10.1371/journal.pbio.1001745}

  \end{itemize}
\end{frame}

% etc
\end{document}
